webextension/background.js

```diff
@@ -0,0 +1,1006 @@
+
+    cumulativeMillisecondsSpentOnDomain(url) {
+      try {
+        const domain = this.findDomainMatch(new URL(url).host);
+        return this._domainsToCheck[domain].totalActiveTime || 0;
+      } catch (_) {
+        return 0;
+      }
+    }
+
+    getDelayBeforePromptingForDomain(url) {
+      const minMillisecondsUserTimeOnDomainBeforePrompt = 0;
+      const maxMillisecondsExtraRandomDelay = 5000;
+
+      const min = Math.max(0, minMillisecondsUserTimeOnDomainBeforePrompt -
+                              this.cumulativeMillisecondsSpentOnDomain(url));
+      const timeout = min + (Math.random() * maxMillisecondsExtraRandomDelay);
+
+      if (this._testingMode) {
+        console.info("Prompting user in", timeout / 1000, "seconds on", url);
+      }
+
+      return timeout;
+    }
+
+    get loaded() {
+      return this._loaded;
+    }
+
+    get testingMode() {
+      return this._testingMode;
+    }
+
+    get reportLanding() {
+      return this._reportLanding;
+    }
+
+    get screenshotFormat() {
+      return {format: "jpeg", quality: 75};
+    }
+
+    get lastPromptTime() {
+      return this._lastPromptTime;
+    }
+
+    get skipPrivateBrowsingTabs() {
+      return this._skipPrivateBrowsingTabs;
+    }
+
+    set skipPrivateBrowsingTabs(bool) {
+      this._skipPrivateBrowsingTabs = bool;
+      this.save({skipPrivateBrowsingTabs: bool});
+    }
+
+    get uiVariant() {
+      return this._uiVariant;
+    }
+
+    set uiVariant(uiVariant) {
+      this._uiVariant = uiVariant;
+      this.save({uiVariant});
+    }
+
+    get appVersion() {
+      return this._appVersion;
+    }
+
+    get buildID() {
+      return this._buildID;
+    }
+
+    get platform() {
+      return this._platform;
+    }
+
+    get releaseChannel() {
+      return this._releaseChannel;
+    }
+  }
+
+  return new Config();
+}());
+
+async function shouldPromptUser(tabId, url) {
+  try {
+    url = new URL(url);
+    return Config.isPromptableURL(url) &&
+           Config.shouldPromptUserNow(url.host) &&
+           (!Config.skipPrivateBrowsingTabs ||
+            !(await browser.tabs.get(tabId)).incognito);
+  } catch (_) {
+    return false;
+  }
+}
+
+function yesOrNo(bool) {
+  return bool ? "yes" : "no";
+}
+
+const portToPageAction = (function() {
+  let port;
+
+  browser.runtime.onConnect.addListener(_port => {
+    // When the page action popup is shown.
+    port = _port;
+    port.onMessage.addListener(onMessageFromPageAction);
+    port.onDisconnect.addListener(function() {
+      // When the page action popup is hidden.
+      port = undefined;
+
+      // Update the page action icon for whichever tab we're on now.
+      TabState.get().then(tabState => {
+        updatePageActionIcon(tabState.tabId);
+      });
+    });
+
+    TabState.get().then(tabState => {
+      tabState.onPageActionShown();
+    });
+  });
+
+  async function send(message) {
+    if (port) {
+      return port.postMessage(message);
+    }
+    console.trace();
```

Comment:
Probably want to remove this or enclose it in some sort of `if (debug)` check for production

Reply:
Sure, I'll do that.